server/Main.hs

```diff
@@ -198,136 +109,110 @@ handleError :: Show e => Either e a -> Scotty.ActionM a
   userId <- case Base64.decodeUnpadded (Text.encodeUtf8 userId') of
     Left err -> fail $ "Failed to base64url decode the user id " <> show userId' <> ": " <> err
     Right res -> pure $ M.UserHandle res
   credentials <- liftIO $
     Database.withTransaction db $ \tx -> do
-      Database.getCredentialsByUserId tx userId
+      Database.getCredentialEntriesByUserId tx userId
   when (null credentials) $ Scotty.raiseStatus HTTP.status404 "User not found"
-  unless
-    (isUnauthenticated session)
-    (Scotty.raiseStatus HTTP.status400 "You need to be unauthenticated to begin login")
-  challenge <- liftIO $ uniformM globalStdGen
-  let options =
-        M.PublicKeyCredentialRequestOptions
-          { M.pkcogRpId = Nothing,
-            M.pkcogTimeout = Nothing,
-            M.pkcogChallenge = challenge,
-            M.pkcogAllowCredentials = map mkCredentialDescriptor credentials,
-            M.pkcogUserVerification = M.UserVerificationRequirementPreferred,
-            M.pkcogExtensions = Nothing
-          }
-  liftIO $ STM.atomically $ casSession sessions sessionId session (Authenticating userId options)
-  Scotty.json $ encodePublicKeyCredentialRequestOptions options
-
-completeLogin :: M.Origin -> M.RpIdHash -> Database.Connection -> TVar Sessions -> Scotty.ActionM ()
-completeLogin origin rpIdHash db sessions = do
-  (sessionId, session) <- getSessionScotty sessions
-  case session of
-    Authenticating userHandle options -> verifyLogin sessionId session userHandle options
-    _ -> Scotty.raiseStatus HTTP.status400 "You need to be authenticating to complete login"
-  where
-    verifyLogin :: SessionId -> Session -> M.UserHandle -> M.PublicKeyCredentialOptions 'M.Get -> Scotty.ActionM ()
-    verifyLogin sessionId session userHandle options = do
-      credential <- Scotty.jsonData @JS.RequestedPublicKeyCredential
-
-      cred <- case decodeRequestedPublicKeyCredential credential of
-        Left err -> fail $ show err
-        Right result -> pure result
-
-      -- TODO: Query for the credential id directly
-      entries <- liftIO $
-        Database.withTransaction db $ \tx -> Database.getCredentialsByUserId tx userHandle
-      entry <- case find ((== M.pkcIdentifier cred) . ceCredentialId) entries of
-        Nothing -> fail "Credential not found"
-        Just entry -> pure entry
-
-      -- step 1 to 17
-      -- We abort if we couldn't attest the credential
-      -- FIXME
-      _newSigCount <- case verifyAssertionResponse origin rpIdHash Nothing entry options cred of
-        Failure (err :| _) -> fail $ show err
-        Success result -> pure result
-      -- FIXME: Set new signature count
-      liftIO $
-        STM.atomically $
-          casSession sessions sessionId session (Authenticated userHandle)
-      Scotty.json @Text "Welcome."
+  options <- liftIO $
+    insertPendingOptions pending $ \challenge -> do
+      M.PublicKeyCredentialRequestOptions
+        { M.pkcogRpId = Nothing,
```

Comment:
Since this server serves as an example, I think it might be nice to explain what certain options do. I.e. that leaving the RpId empty results it in being equivalent to the domain.

Although this can ofc also be  done in the followup PR for the server rework.

Reply:
Yeah this sounds better for another PR. These lines weren't changed in this one (which is more obvious if you turn on "Hide Whitespace" in the GitHub's diff settings)